packages/blade/src/components/SideNav/SideNav.tsx

```diff
@@ -0,0 +1,233 @@
+      },
+    },
+  };
+});
+
+const getL1MenuClassName = ({
+  isL1Collapsed,
+  isL1Hovered,
+  isTransitioning,
+}: {
+  isL1Collapsed: boolean;
+  isL1Hovered: boolean;
+  isTransitioning: boolean;
+}): string => {
+  const isMenuCollapsed = isL1Collapsed && !isL1Hovered;
+
+  if (isMenuCollapsed) {
+    if (isTransitioning) {
+      return `${COLLAPSED} ${TRANSITIONING}`;
+    }
+
+    return COLLAPSED;
+  }
+
+  return '';
+};
+
+const SideNav = ({ children, isOpen, onDismiss }: SideNavProps): React.ReactElement => {
+  const l2PortalContainerRef = React.useRef(null);
+  const l1ContainerRef = React.useRef<HTMLDivElement>(null);
+  const [isL1Collapsed, setIsL1Collapsed] = React.useState(false);
+  const [isMobileL2Open, setIsMobileL2Open] = React.useState(false);
+  const [isL1Hovered, setIsL1Hovered] = React.useState(false);
+  const [isHoverAgainEnabled, setIsHoverAgainEnabled] = React.useState(true);
+  const [isTransitioning, setIsTransitioning] = React.useState(false);
+  const [l2DrawerTitle, setL2DrawerTitle] = React.useState('');
+
+  console.count('SideNav');
+  const isMobile = useIsMobile();
+
+  const closeMobileNav = (): void => {
+    if (isMobile) {
+      setIsMobileL2Open(false);
+      onDismiss?.();
+    }
+  };
+
+  const footerChild: React.ReactNode[] = [];
+  const restChild: React.ReactNode[] = [];
+
+  React.Children.map(children, (child) => {
+    if (getComponentId(child) === 'SideNavFooter') {
+      footerChild.push(child);
+    } else {
+      restChild.push(child);
+    }
+  });
+
+  const onLinkActiveChange: SideNavContextType['onLinkActiveChange'] = (args) => {
+    if (args.level === 1 && args.isL2Trigger && args.isActive) {
+      if (isMobile) {
+        setL2DrawerTitle(args.title);
+        if (!args.isFirstRender) {
+          // we don't want to open sidenav in mobile in first render
+          setIsMobileL2Open(true);
+        }
+        return;
+      }
+
+      setIsL1Collapsed(true);
+
+      if (!args.isFirstRender) {
+        setIsTransitioning(true);
+        setIsL1Hovered(false);
+        setIsHoverAgainEnabled(false);
+        // For some delay, we disable hover to expand behaviour to avoid buggy flicker when cursor is on L1 while its trying to close
+        setTimeout(() => {
+          setIsHoverAgainEnabled(true);
+        }, 500);
+      }
+    }
+
+    if (args.level === 1 && !args.isL2Trigger && args.isActive) {
+      if (isMobile) {
+        setIsMobileL2Open(false);
+        return;
+      }
+      setIsL1Collapsed(false);
+    }
+  };
+
+  const contextValue = React.useMemo(
+    () => ({
+      l2PortalContainerRef,
+      onLinkActiveChange,
+      closeMobileNav,
+      isL1Collapsed: isMobile ? isMobileL2Open : isL1Collapsed,
+      setIsL1Collapsed,
+    }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [isL1Collapsed, isMobile, isMobileL2Open],
+  );
+
+  return (
+    <SideNavContext.Provider value={contextValue}>
+      {isMobile && onDismiss ? (
```

Comment:
remove onDismiss check.

Reply:
let it be no it makes typescript happy. Will have to put null assertion below otherwise 🙈